### PR TITLE
[Diagnostics] Fix all of the edge cases and remove `diagnoseContextualConversionError`

### PR DESCRIPTION
**Pull request by xedin**
* Diagnose an attempt to init/use dictionary with array literal in CSGen
* ForEach: Don't diagnose unrelated errors if type doesn't conform to Sequence
* Tidy up restriction handling in contextual failures
* Let `try?` infer type from sub-expression even with contextual mismatch
* Let optional chaining infer a type from sub-expression even with contextual mismatch
* Detect missing/extraneous arguments even if there is optional injection required
* Account for nested structural errors related to function argument mismatches

### CODE REVIEW

**Comment by xedin** [comment]
Now only `FailureDiagnosis::visitApplyExpr` remains...

**Comment by xedin** [comment]
@swift-ci please smoke test

File: lib/Sema/CSFix.cpp [hunk 1]
```diff
@@ -262,11 +262,15 @@ ContextualMismatch *ContextualMismatch::create(ConstraintSystem &cs, Type lhs,
...
-      return std::make_tuple(cs.getContextualTypePurpose(locator->getAnchor()),
-                             cs.getType(anchor),
-                             contextualType);
+  if (locator->findLast<LocatorPathElt::ContextualType>()) {
+    assert(locator->isLastElement<LocatorPathElt::ContextualType>() ||
+           locator->isLastElement<LocatorPathElt::FunctionArgument>());
+
+    auto *anchor = locator->getAnchor();
+    auto contextualType = cs.getContextualType(anchor);
+    auto exprType = cs.getType(locator->getAnchor());
```
Comment (LucianoPAlmeida) [inline lib/Sema/CSFix.cpp]: The anchor from the line above could be reused instead of `locator->getAnchor()`?

Comment (xedin) [reply lib/Sema/CSFix.cpp]: Thank you!

Comment (xedin) [reply lib/Sema/CSFix.cpp]: Fixed :)

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please smoke test

**Review by hborla — APPROVED** [review]

**Comment by xedin** [comment]
@swift-ci please smoke test Linux platform

**Pull request merged**